File: 17_Math&Geometry/01_happy_number/main.py
def sum_of_squares(n):
    sum = 0
    while n > 0:
        digit = n % 10
        sum += digit * digit
        n //= 10
    return sum

def is_happy(n):
    seen_numbers = set()
    while n != 1 and n not in seen_numbers:
        seen_numbers.add(n)
        n = sum_of_squares(n)
    return n == 1

# Testing the function
test1 = 100
test2 = 101

print(f"Is {test1} a happy number? {'True' if is_happy(test1) else 'False'}")
print(f"Is {test2} a happy number? {'True' if is_happy(test2) else 'False'}")

File: 17_Math&Geometry/01_happy_number/main.cpp
#include <iostream>
#include <unordered_set>

class Solution {
public:
    bool isHappy(int n) {
        std::unordered_set<int> seenNumbers;
        while (n != 1 && seenNumbers.find(n) == seenNumbers.end()) {
            seenNumbers.insert(n);
            n = sumOfSquares(n);
        }
        return n == 1;
    }

private:
    int sumOfSquares(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }
};

int main() {
    Solution solution;
    int test1 = 100;
    int test2 = 101;

    std::cout << "Is " << test1 << " a happy number? " << (solution.isHappy(test1) ? "True" : "False") << std::endl;
    std::cout << "Is " << test2 << " a happy number? " << (solution.isHappy(test2) ? "True" : "False") << std::endl;

    return 0;
}
